File: W6/py/39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#

# @lc code=start


class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        results = []
        self.helper([], 0, results, target, sorted(candidates))
        return results

    def helper(self, current, index, results, target, nums):
        if sum(current) > target:
            return
        if sum(current) == target:
            results.append(list(current))
            return

        for i in range(index, len(nums)):
            current.append(nums[i])
            # if you don't want to access youself again, do i + 1
            self.helper(current, i, results, target, nums)
            current.pop()
# @lc code=end

File: swift/39.combination-sum.swift
class Solution {
    internal var res:[[Int]] = []
    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        helper(candidates.sorted(), target, 0, 0, [])
        return res
    }
    
    func helper(_ candidates: [Int], _ target: Int, _ index: Int, _ sum: Int, _ curr: [Int]) {
        let length = candidates.count
        
        if sum == target {
            res += [curr]
            return
        }
        
        for i in index-1..<length {
            if i == -1 { continue }
            
            let item = candidates[i]
            let newSum = sum + item
            if newSum > target {
                break
            }
            
            helper(candidates, target, i + 1, newSum, curr + [item])
        }
        
    }
}